staging/src/k8s.io/component-base/config/testing/helpers.go

```diff
@@ -0,0 +1,78 @@
+package testing
+
+import (
+	"bytes"
+	"io/ioutil"
+	"os"
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+)
+
+func RunTestsOnYAMLData(t *testing.T, scheme *runtime.Scheme, tests []TestCase, codecs serializer.CodecFactory) {
+	for _, rt := range tests {
+		t.Run(rt.name, func(t *testing.T) {
+			obj, err := decodeTestData(rt.in, scheme, rt.inGVK, codecs)
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			const mediaType = runtime.ContentTypeYAML
+			info, ok := runtime.SerializerInfoForMediaType(codecs.SupportedMediaTypes(), mediaType)
+			if !ok {
+				t.Errorf("unable to locate encoder -- %q is not a supported media type", mediaType)
+			}
+
+			encoder := codecs.EncoderForVersion(info.Serializer, rt.outGV)
+
+			actual, err := runtime.Encode(encoder, obj)
+			if err != nil {
+				t.Fatal(err)
+			}
+
+			expected, err := ioutil.ReadFile(rt.out)
+			if err != nil && !os.IsNotExist(err) {
+				t.Fatalf("couldn't read test data: %v", err)
+			}
+
+			needsUpdate := false
+			if os.IsNotExist(err) {
+				needsUpdate = true
+				t.Error("couldn't find test data")
+			} else {
+				if !bytes.Equal(expected, actual) {
+					t.Errorf("the expected and actual output differs.\n\tin: %s\n\tout: %s\n\tgroupversion: %s\n\tdiff: \n%s\n",
```

Comment:
It can be helpful in a diff error to include the direction (what - or + means):

```go
t.Errorf("Output does not match expected, diff (- want, + got):\n%s\n", ...)
```
If we're providing a diff, it may be too large an error to include the raw input data as well (and the diff effectively expresses what this is).